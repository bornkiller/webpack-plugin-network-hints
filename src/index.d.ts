export interface NetworkHintsOptions {
  prefetch?: string[];
  preload?: string[];
}

export class NetworkHintsPlugin {
  constructor(options: NetworkHintsOptions) {}
}
