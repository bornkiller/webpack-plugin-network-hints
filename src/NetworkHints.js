/**
 * @description - webpack plugin for inject network hints
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import { resolve } from 'url';
import { match } from 'minimatch';
import HtmlWebpackPlugin from 'html-webpack-plugin';

// internal
import RequireHtmlWebpackPluginError from './exceptions/RequireHtmlWebpackPluginError';
import UnsupportHtmlWebpackPluginVersionError from './exceptions/UnsupportHtmlWebpackPluginVersionError';
import preloadAttrAlgorithm from './preload';

// scope
const defaultOptions = {
  prefetch: [],
  preload: [],
};

class NetworkHints {
  constructor(options) {
    this.options = { ...defaultOptions, ...options };
  }

  apply(compiler) {
    compiler.hooks.compilation.tap(
      'WebpackPluginNetworkHints',
      (compilation) => {
        const [HtmlWebpackPluginInstance] = compiler.options.plugins.filter(
          (plugin) => plugin.constructor.name === 'HtmlWebpackPlugin'
        );

        if (!HtmlWebpackPluginInstance) {
          compilation.errors.push(new RequireHtmlWebpackPluginError());
        }
        // pre-v4 HtmlWebpackPlugin
        else if (
          Reflect.has(compilation.hooks, 'htmlWebpackPluginAlterAssetTags')
        ) {
          compilation.errors.push(new UnsupportHtmlWebpackPluginVersionError());
        } else {
          // get available hooks directly
          const hooks = HtmlWebpackPlugin.getHooks(compilation);

          hooks.alterAssetTagGroups.tapAsync(
            'InjectExternalPlugin',
            (assetTagGroups, callback) => {
              const { publicPath } = compiler.options.output;
              const assets = Reflect.ownKeys(compilation.assets);
              const prefetch = this.options.prefetch
                .reduce(
                  (acc, pattern) => [
                    ...acc,
                    ...match(assets, pattern, { matchBase: true }),
                  ],
                  []
                )
                .sort()
                .map((asset) => resolve(publicPath, asset))
                .map((asset) =>
                  HtmlWebpackPlugin.createHtmlTagObject('link', {
                    rel: 'prefetch',
                    href: asset,
                  })
                );
              const preload = this.options.preload
                .reduce(
                  (acc, pattern) => [
                    ...acc,
                    ...match(assets, pattern, { matchBase: true }),
                  ],
                  []
                )
                .sort() // consistant assets order
                .map((asset) => resolve(publicPath, asset))
                .map((asset) =>
                  HtmlWebpackPlugin.createHtmlTagObject(
                    'link',
                    preloadAttrAlgorithm(asset)
                  )
                );

              const payload = {
                ...assetTagGroups,
                headTags: [...assetTagGroups.headTags, ...prefetch, ...preload],
              };

              callback(null, payload);
            }
          );
        }
      }
    );
  }
}

export default NetworkHints;
