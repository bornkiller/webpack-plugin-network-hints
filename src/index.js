/**
 * @description - nothing but re-export
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// eslint-disable-next-line import/prefer-default-export
export { default as NetworkHintsPlugin } from './NetworkHints';
