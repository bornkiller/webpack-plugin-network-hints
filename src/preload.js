/**
 * @description - calculate preload link attributes
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import { extname } from 'path';
import { lookup } from 'mime-types';

// scope
const scripts = ['.js', '.mjs'];
const stylesheets = ['.css'];
const fonts = ['.woff', '.woff2', '.ttf', '.eot'];
const images = ['.jpg', '.jpeg', '.png', '.gif', '.webp'];
const documents = ['.html'];

/**
 * @description - compose specific attrs for specific files
 *
 * @param {string} asset
 */
function preloadSpecificAttrs(asset) {
  const extension = extname(asset);

  switch (true) {
    case scripts.includes(extension):
      return { as: 'script' };
    case stylesheets.includes(extension):
      return { as: 'stylesheet', onload: `this.rel='stylesheet'` };
    case fonts.includes(extension):
      return { as: 'font', crossorigin: 'anonymous' };
    case images.includes(extension):
      return { as: 'image' };
    case documents.includes(extension):
      return { as: 'document' };
    default:
      return {};
  }
}

/**
 * @param {string} asset
 *
 * @return {object}
 */
export default function preload(asset) {
  const attrs = {
    rel: 'preload',
    href: asset,
    type: lookup(asset),
  };
  const specifics = preloadSpecificAttrs(asset);

  return { ...attrs, ...specifics };
}
