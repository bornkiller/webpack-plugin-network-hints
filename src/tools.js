/**
 * @description - tool method
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

/* eslint-disable import/prefer-default-export */

/**
 * @description - create prefetch tags
 *
 * @param {string} asset
 */
export function createPrefetchTags(asset) {
  return {
    tagName: 'link',
    selfClosingTag: false,
    voidTag: true,
    attributes: {
      rel: 'prefetch',
      href: asset,
    },
  };
}

/**
 * @description - create preload tags
 *
 * @param {string} asset
 * @param {function} algorithm - calculate preload additional attributes
 */
export function createPreloadTags(asset, algorithm) {
  return {
    tagName: 'link',
    selfClosingTag: false,
    voidTag: true,
    attributes: algorithm(asset),
  };
}
