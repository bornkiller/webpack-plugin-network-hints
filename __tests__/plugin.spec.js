/* eslint-disable import/no-extraneous-dependencies, no-console */

// packages
import fs from 'fs';
import { resolve } from 'path';
import webpack from 'webpack';
import MemoryFS from 'memory-fs';
import HtmlWebpackPlugin from 'html-webpack-plugin';

// internal
import { NetworkHintsPlugin } from '../src';

// Scope
const mfs = Reflect.construct(MemoryFS, []);

describe('plugin test suits', () => {
  it('should complete standard workflow', (done) => {
    const configuration = {
      entry: resolve(__dirname, '__fixture__', 'index.js'),
      resolve: {
        extensions: ['.js', '.jpg'],
      },
      output: {
        path: resolve(process.cwd(), 'dist'),
        filename: '[name].js',
        publicPath: '/',
      },
      module: {
        rules: [
          {
            test: /\.jpg$/,
            use: [
              {
                loader: 'file-loader',
                options: {
                  name: 'static/media/[name].[hash:8].[ext]',
                },
              },
            ],
          },
        ],
      },
      plugins: [
        new HtmlWebpackPlugin({
          template: resolve(__dirname, '__fixture__', 'index.html'),
          inject: 'body',
        }),
        new NetworkHintsPlugin({
          prefetch: ['*.jpg'],
        }),
      ],
    };

    const compiler = webpack(configuration);
    const outputPath = `${configuration.output.path}/index.html`;

    compiler.inputFileSystem = fs;
    compiler.outputFileSystem = mfs;

    compiler.run((err) => {
      try {
        const content = mfs.readFileSync(outputPath, 'utf8');

        expect(err).toBeNull();
        expect(content).toMatchSnapshot();
        done();
      } catch (error) {
        done(error);
      }
    });
  });
});
