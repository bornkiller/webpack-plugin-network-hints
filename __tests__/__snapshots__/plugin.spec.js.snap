// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin test suits should complete standard workflow --> prefetch 1`] = `"<!doctype html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"/><title>Jest</title><link rel=\\"prefetch\\" href=\\"/static/media/david-marcu-114194-unsplash.e5c27cdd.jpg\\"><link rel=\\"prefetch\\" href=\\"/static/media/eberhard-grossgasteiger-1482753-unsplash.63c69b5c.jpg\\"><link rel=\\"prefetch\\" href=\\"/static/media/fezbot2000-278419-unsplash.feec5470.jpg\\"></head><body><script src=\\"/main.js\\"></script></body></html>"`;

exports[`plugin test suits should complete standard workflow --> preload 1`] = `"<!doctype html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"/><title>Jest</title><link rel=\\"preload\\" href=\\"/static/media/david-marcu-114194-unsplash.e5c27cdd.jpg\\" type=\\"image/jpeg\\" as=\\"image\\"><link rel=\\"preload\\" href=\\"/static/media/eberhard-grossgasteiger-1482753-unsplash.63c69b5c.jpg\\" type=\\"image/jpeg\\" as=\\"image\\"><link rel=\\"preload\\" href=\\"/static/media/fezbot2000-278419-unsplash.feec5470.jpg\\" type=\\"image/jpeg\\" as=\\"image\\"></head><body><script src=\\"/main.js\\"></script></body></html>"`;
